src/main/cpp/src/zorder.cu

```diff
@@ -27,6 +27,118 @@
 #include <thrust/iterator/constant_iterator.h>
 #include <thrust/iterator/counting_iterator.h>
 
+namespace {
+
+// pretends to be an array of uint32_t, but really only stores
+// the data in a long with a set number of bits allocated for
+// each item
+struct long_backed_array {
+  long_backed_array() = delete;
+  ~long_backed_array() = default;
+  long_backed_array(long_backed_array const&) = default;  ///< Copy constructor
+  long_backed_array(long_backed_array&&) = default;  ///< Move constructor
+  inline __device__ explicit long_backed_array(int32_t num_bits): data(0), 
+    num_bits(num_bits),  mask(static_cast<uint64_t>((1L << num_bits) - 1)) {}
```

Comment:
Should `num_bits` be named something more like `num_bits_per_entry` or some such? It took me a while to make the connection that it wasn't the total number of bits in the array. I also don't see any verification that the number of bits desired per element * number of elements is < sizeof(long).